pairing/src/lib.rs

```diff
@@ -115,3 +85,33 @@ pub trait PairingCurveAffine: CurveAffine {
     /// Perform a pairing
     fn pairing_with(&self, other: &Self::Pair) -> Self::PairingResult;
 }
+
+/// An engine that can compute sums of pairings in an efficient way.
+pub trait MultiMillerLoop: Engine {
+    /// The type returned by `Engine::miller_loop`.
+    type Result: MillerLoopResult<Gt = Self::Gt>;
+
+    /// Computes $$\sum_{i=1}^n \textbf{ML}(a_i, b_i)$$ given a series of terms
+    /// $$(a_1, b_1), (a_2, b_2), ..., (a_n, b_n).$$
+    fn multi_miller_loop(
+        terms: &[(
+            &Self::G1Affine,
+            &<Self::G2Affine as PairingCurveAffine>::Prepared,
+        )],
+    ) -> Self::Result;
+}
+
+/// Represents results of a Miller loop, one of the most expensive portions of the pairing
+/// function.
+///
+/// `MillerLoopResult`s cannot be compared with each other until
+/// [`MillerLoopResult::final_exponentiation`] is called, which is also expensive.
```

Comment:
This comment suggests that calling `final_exponentiation` is the preferred way to compare two `MillerLoopResult`s. Even if you can't rewrite to a single `multi_miller_loop`, you can still conjugate one of the loop results, multiply by the other and compare to the identity in `Gt`. So it might be worth exposing a `std::cmp::Eq` implemention that does that.

Reply:
That would make equality-checking very expensive, which would be very non-obvious in `a == b` usage.

Reply:
I think that's fine as long as it's documented. But it doesn't block this PR, anyway.